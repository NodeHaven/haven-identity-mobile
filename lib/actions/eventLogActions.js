// Copyright (C) 2018 ConsenSys AG
//
// This file is part of uPort Mobile App.
//
// uPort Mobile App is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// uPort Mobile App is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with uPort Mobile App.  If not, see <http://www.gnu.org/licenses/>.
//
import { STORE_LAST_BLOCK_FOR_LOGS, STORE_EVENTS } from '../constants/EventLogActionTypes'

// ADDING NEW ACTIONS PLEASE READ THIS!!!!
//
// If the action should be backed up add the `, _backup` attribute to the action.
// When should it be backed up?
//
// - is it backed by the reducer and not a saga?
// - is it something that can not be refreshed from the block chain etc? (eg. no need to backup nonces and balances)
// - is it not just related to UX? Some reducers are temporary and would not make sense on multiple device

const _backup = true

export function storeLastBlockForLogs (contract, block) {
  return {
    type: STORE_LAST_BLOCK_FOR_LOGS,
    contract,
    block,
    _backup
  }
}

export function storeEvents (contract, events) {
  return {
    type: STORE_EVENTS,
    contract,
    events,
    _backup
  }
}
