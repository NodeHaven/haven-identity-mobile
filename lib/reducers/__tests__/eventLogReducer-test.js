// Copyright (C) 2018 ConsenSys AG
//
// This file is part of uPort Mobile App.
//
// uPort Mobile App is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// uPort Mobile App is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with uPort Mobile App.  If not, see <http://www.gnu.org/licenses/>.
//
import {storeLastBlockForLogs, storeEvents} from 'uPortMobile/lib/actions/eventLogActions.js'
import baseReducer from 'uPortMobile/lib/reducers/eventLogReducer.js'
import { toClj, toJs, hashMap } from 'mori'

const reducer = (state, action) => toJs(baseReducer(state, action))

const empty = hashMap()

const withEvents = toClj({
  '0x010102': {
    block: 10000,
    events: [{type: 'test'}]
  },
  '0x010101': {
    block: 10000,
    events: [{type: 'test'}]
  }
})

it('creates a STORE_LAST_BLOCK_FOR_LOGS action', () => {
  expect(reducer(null, storeLastBlockForLogs('0x010102', 12313))).toMatchSnapshot()
  expect(reducer(withEvents, storeLastBlockForLogs('0x010102', 12313))).toMatchSnapshot()
})

it('creates a STORE_EVENTS action', () => {
  expect(reducer(null, storeEvents('0x010102', [{type: 'something'}]))).toMatchSnapshot()
  expect(reducer(withEvents, storeEvents('0x010102', [{type: 'something'}]))).toMatchSnapshot()
})

it('ignores unsupported action', () => {
  expect(baseReducer(undefined, {type: 'UNSUPPORTED'})).toEqual(empty)
  expect(baseReducer(empty, {type: 'UNSUPPORTED'})).toEqual(empty)
  expect(baseReducer(withEvents, {type: 'UNSUPPORTED'})).toEqual(withEvents)
})
