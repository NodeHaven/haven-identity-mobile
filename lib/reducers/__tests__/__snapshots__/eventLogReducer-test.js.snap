// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates a STORE_EVENTS action 1`] = `
Object {
  "0x010102": Object {
    "events": Array [
      Object {
        "type": "something",
      },
    ],
  },
}
`;

exports[`creates a STORE_EVENTS action 2`] = `
Object {
  "0x010101": Object {
    "block": 10000,
    "events": Array [
      Object {
        "type": "test",
      },
    ],
  },
  "0x010102": Object {
    "block": 10000,
    "events": Array [
      Object {
        "type": "test",
      },
      Object {
        "type": "something",
      },
    ],
  },
}
`;

exports[`creates a STORE_LAST_BLOCK_FOR_LOGS action 1`] = `
Object {
  "0x010102": Object {
    "block": "3019",
  },
}
`;

exports[`creates a STORE_LAST_BLOCK_FOR_LOGS action 2`] = `
Object {
  "0x010101": Object {
    "block": 10000,
    "events": Array [
      Object {
        "type": "test",
      },
    ],
  },
  "0x010102": Object {
    "block": "3019",
    "events": Array [
      Object {
        "type": "test",
      },
    ],
  },
}
`;
